runtime/gc_vlhgc/VLHGCAccessBarrier.cpp

```diff
@@ -272,34 +272,63 @@ MM_VLHGCAccessBarrier::jniGetPrimitiveArrayCritical(J9VMThread* vmThread, jarray
 #endif
 	}
 
-	if (shouldCopy) {
-		GC_ArrayObjectModel* indexableObjectModel = &_extensions->indexableObjectModel;
-		I_32 sizeInElements = (I_32)indexableObjectModel->getSizeInElements(arrayObject);
-		UDATA sizeInBytes = indexableObjectModel->getDataSizeInBytes(arrayObject);
-		data = functions->jniArrayAllocateMemoryFromThread(vmThread, sizeInBytes);
-		if(NULL == data) {
-			functions->setNativeOutOfMemoryError(vmThread, 0, 0);	// better error message here?
-		} else {
-			indexableObjectModel->memcpyFromArray(data, arrayObject, 0, sizeInElements);
-			if(NULL != isCopy) {
-				*isCopy = JNI_TRUE;
+#if defined(LINUX)
+#if defined(J9VM_GC_ENABLE_DOUBLE_MAP)
+	bool successDoubleMap = false;
+        if(shouldCopy && _extensions->indexableObjectModel.isDoubleMappingEnabled()) {
+	
+		GC_HashTableIterator hashTableIterator(_extensions->getArrayletHashTable());
+                ArrayletTableEntry *slot = NULL;
+                bool found = false;
+                while(NULL != (slot = (ArrayletTableEntry *)hashTableIterator.nextSlot())) {
+			if(slot->heapAddr == (void *)arrayObject) {
+				found = true;
+				data = slot->contiguousAddr;
+				break;
 			}
 		}
-		vmThread->jniCriticalCopyCount += 1;
-	} else {
-		// acquire access and return a direct pointer
-		MM_JNICriticalRegion::enterCriticalRegion(vmThread, true);
-		Assert_MM_true(vmThread->publicFlags & J9_PUBLIC_FLAGS_VM_ACCESS);
-		arrayObject = (J9IndexableObject*)J9_JNI_UNWRAP_REFERENCE(array);
-		data = (void *)_extensions->indexableObjectModel.getDataPointerForContiguous(arrayObject);
-		if(NULL != isCopy) {
-			*isCopy = JNI_FALSE;
+
+		if(data == NULL || !found) {
+			successDoubleMap = false;
+		} else {
+			successDoubleMap = true;
 		}
+
+		_successDoubleMap = successDoubleMap;
+        }
+	if(!successDoubleMap) 
+#endif /* J9VM_GC_ENABLE_DOUBLE_MAP */
+#endif /* LINUX */
+	{
+		if (shouldCopy) {
+			GC_ArrayObjectModel* indexableObjectModel = &_extensions->indexableObjectModel;
+			I_32 sizeInElements = (I_32)indexableObjectModel->getSizeInElements(arrayObject);
+			UDATA sizeInBytes = indexableObjectModel->getDataSizeInBytes(arrayObject);
+			data = functions->jniArrayAllocateMemoryFromThread(vmThread, sizeInBytes);
+			if(NULL == data) {
+				functions->setNativeOutOfMemoryError(vmThread, 0, 0);	// better error message here?
```

Comment:
Allowing this to fallthrough and modify the jniCriticalCopyCount seems odd - shouldn't it avoid modifying those counts?